Add notes on how to connect Keycloak and Vault (#24)
projectsyn/component-vault: docs/modules/ROOT/pages/tutorials/install.adoc

This means in practice you will want to put the production vault in a separate namespace.

More information on how component instantiation works can be found https://syn.tools/commodore/reference/architecture.html#_component_instantiation[here].

== OIDC Integration

Vault has the option to authenticate users through OIDC.

> This guide describes the manual setup of the Vault OIDC integration for a Lieutenant connected Vault.

Reviewer: ```suggestion
This guide describes the manual setup of the Vault OIDC integration for a Vault instance which is associated with a Lieutenant instance.
```

[WARNING]
====
It should also be possible to configure OIDC integration through the component.

> However, this requires a client secret we don't recommend doing so when integrating Vault with Lieutenant as this will introduce a dependency for Vault on itself making setup and disaster recovery more difficult.

Reviewer: ```suggestion
However, this requires a client secret which should be stored in Vault.
we don't recommend configuring OIDC through the component when integrating Vault with Lieutenant as this will introduce a dependency for Vault on itself making setup and disaster recovery more difficult.
```

====

=== Configure OIDC Provider

[NOTE]
====

> This guide assumes Keycloak to be the OIDC provider, but other OIDC provider should also work.

Reviewer: ```suggestion
This guide assumes that the OIDC provider is a Keycloak instance, but other OIDC providers should also work.
```

. In the appropriate realm, create a OIDC client with:
+
[source]
-----
Access Type: Confidential
Redirect URLs:
  - https://${VAULT_URL}/ui/vault/auth/oidc/oidc/callback"
  - https://${VAULT_URL}/oidc/callback"
-----

. Add a role `syn-root` under the `Roles` tab in the OIDC client

. Under the `Mappers` tab add a `User Client Role` mapper.
Choose the created OIDC client as the `Client ID` and set the `Token Claim Name` to `groups`.

. Open the Keycloak group or user that should get access to Vault and add them to the `syn-root` role of the created OIDC client.

. Copy client secret of the created OIDC client located in the `Credentials` tab of the client.

=== Configure Vault

. Add a new OIDC Authenication Method in the Vault UI under the Access tab with the following configuration
+
[source]
-----
List method when unauthenticated: true
OIDC discovery URI: https://${KEYCLOAK_URL}/auth/realms/${REALM}
Default role: default
OIDC client ID: ${CLIENT_ID}
OIDC client Secret: ${CLIENT_SECRET}
-----
+
At this point you should be able to login to Vault using OIDC, but after the login you shouldn't be able to access anything but the `cubbyhole`.

> . Open a CLI connection to Vault, you can do this by clicking on the CLI icon in the UI, an enter

Reviewer: ```suggestion
. Open a CLI connection to Vault, you can do this by clicking on the CLI icon in the UI, and enter
```
